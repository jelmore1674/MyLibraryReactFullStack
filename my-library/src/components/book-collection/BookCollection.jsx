import { React, useEffect, useState } from 'react';
import BookCard from '../book-card/BookCard';
import { getLibrary } from './list';

export default function BookCollection({
	user,
	changedLibrary,
	setChangedLibrary,
}) {
	const [library, setLibrary] = useState([]);

	// useEffect is used to fetch the library collection
	useEffect(() => {
		async function fetchLibrary() {
			const books = await getLibrary(
				`${process.env.REACT_APP_HOST}/library-item/${user.userid}`
			);
			setLibrary((prevState) => books);
		}
		fetchLibrary();
	}, [changedLibrary, user.userid]);

	const removeBook = async (id) => {
		try {
			const response = await fetch(
				`${process.env.REACT_APP_HOST}/library-item`,
				{
					method: 'delete',
					headers: {
						'Access-Control-Allow-Origin': '*',
						'Content-Type': 'application/json',
					},
					body: JSON.stringify({
						id: id,
						remove: true,
					}),
				}
			);
			const data = await response.json();
			if (data === 'success') {
				setChangedLibrary((prevStatus) => prevStatus + 1);
			} else {
				Promise.reject('could not remove book from library');
			}
		} catch (err) {
			console.log(err);
		}
	};

	const updateBook = async (id) => {
		try {
			for (var i = 0; i < library.length; i++) {
				if (library[i].id === id) {
					const completedStatus = !library[i].completed;
					const response = await fetch(
						`${process.env.REACT_APP_HOST}/library-item`,
						{
							method: 'put',
							headers: {
								'Access-Control-Allow-Origin': '*',
								'Content-Type': 'application/json',
							},
							body: JSON.stringify({
								id: id,
								update: true,
								completed: completedStatus,
							}),
						}
					);
					const data = await response.json();
					if (data === 'success') {
						setChangedLibrary((prevStatus) => ++prevStatus);
					} else return Promise.reject(data);
				}
			}
		} catch (err) {
			console.log(err);
		}
	};

	function createBookCard(book) {
		return (
			<BookCard
				key={book.id}
				id={book.id}
				title={book.title}
				author={book.author}
				pages={book.pages}
				finished={book.completed}
				removeBook={removeBook}
				updateBook={updateBook}
			/>
		);
	}

	return (
		<div className='container mt-5'>
			<div
				data-testid='bookCollection'
				className='row rol-cols-xs-1 row-cols-sm-2 row-cols-md-2 row-cols-lg-3 row-cols-xl-4 gy-4 justify-content-center align-items-center'>
				{library.map(createBookCard)}
			</div>
		</div>
	);
}
