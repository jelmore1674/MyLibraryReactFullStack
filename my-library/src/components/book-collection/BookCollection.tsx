import { useEffect } from 'react';

import { useDispatch, useSelector } from 'react-redux';
import { Book, fetchLibraryByUserId } from '../../redux/library/librarySlice';
import { RootState } from '../../redux/store';

import BookCard from '../book-card/BookCard';

export default function BookCollection(): JSX.Element {
	const { library } = useSelector((state: RootState) => state.library);
	const { user } = useSelector((state: RootState) => state.user);
	const dispatch = useDispatch();

	// useEffect is used to fetch the library collection
	useEffect(() => {
		if (user) {
			dispatch(fetchLibraryByUserId(user.userid));
		}
	}, [dispatch, user]);

	const removeBook = async (id: number): Promise<void> => {
		try {
			const response = await fetch(`/api/library-item`, {
				method: 'delete',
				headers: {
					'Access-Control-Allow-Origin': '*',
					'Content-Type': 'application/json',
				},
				body: JSON.stringify({
					id: id,
					remove: true,
				}),
			});
			const data = await response.json();
			if (data === 'success') {
				if (user) {
					dispatch(fetchLibraryByUserId(user.userid));
				}
			} else {
				throw new Error('could not remove book from library');
			}
		} catch (err) {
			console.error(err);
		}
	};

	const updateBook = async (id: number): Promise<void> => {
		try {
			for (var i = 0; i < library.length; i++) {
				if (library[i].id === id) {
					const completedStatus = !library[i].completed;
					const response = await fetch(`/api/library-item`, {
						method: 'put',
						headers: {
							'Access-Control-Allow-Origin': '*',
							'Content-Type': 'application/json',
						},
						body: JSON.stringify({
							id: id,
							update: true,
							completed: completedStatus,
						}),
					});
					const data = await response.json();
					if (data === 'Success') {
						if (user) {
							dispatch(fetchLibraryByUserId(user.userid));
						}
					} else {
						throw new Error('could not update book in library');
					}
				}
			}
		} catch (err) {
			console.error(err);
		}
	};

	function createBookCard(book: Book): JSX.Element | undefined {
		if (book.id) {
			return (
				<BookCard
					key={book.id}
					id={book.id}
					title={book.title}
					author={book.author}
					pages={book.pages}
					finished={book.completed}
					removeBook={removeBook}
					updateBook={updateBook}
				/>
			);
		}
	}

	return (
		<div className='container mt-5'>
			<div
				data-testid='bookCollection'
				className='row rol-cols-xs-1 row-cols-sm-2 row-cols-md-2 row-cols-lg-3 row-cols-xl-4 gy-4 justify-content-center align-items-center'>
				{library.map(createBookCard)}
			</div>
		</div>
	);
}
