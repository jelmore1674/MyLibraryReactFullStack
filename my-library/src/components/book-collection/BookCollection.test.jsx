import { render, screen } from '@testing-library/react';

import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import BookCollection from './BookCollection';

describe('BookCollection', () => {
	const user = { userid: 64, email: 'demo@demo.com', name: 'Demo' };
	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(<BookCollection user={user} />, div);
	});

	it('should render Books', async () => {
		const { getByTestId } = render(<BookCollection user={user} />);
		await screen.findAllByTestId('bookCard');
		const bookCard = screen.queryAllByTestId('bookCard');
		expect(bookCard).toBeTruthy();
	});
	it('should match snapshot', async () => {
		const tree = renderer
			.create(await (<BookCollection user={user} />))
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
