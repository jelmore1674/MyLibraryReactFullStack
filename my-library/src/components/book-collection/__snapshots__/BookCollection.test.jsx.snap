// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookCollection should match snapshot 1`] = `
<div
  className="container mt-5"
>
  <div
    className="row rol-cols-xs-1 row-cols-sm-2 row-cols-md-2 row-cols-lg-3 row-cols-xl-4 gy-4 justify-content-center align-items-center"
    data-testid="bookCollection"
  />
</div>
`;
